[Retiarii] refactor of NAS doc

QuanluZhang opened the pull request:
- [x] promote Retiarii framework as default NAS framework on NNI
- [x] remove documents of NNI NAS 1.0 (added reference to NNI NAS 1.0 of v2.2)
- [x] make pure-python execution engine as default execution engine for Retiarii

ultmaster commented on docs/en_US/NAS/ExecutionEngines.rst:
```diff
@@ -0,0 +1,48 @@
+2. Add ``config.execution_engine = 'py'`` to ``RetiariiExeConfig``.
+3. If you need to export top models, formatter needs to be set to ``dict``. Exporting ``code`` won't work with this engine.
+
+.. note:: You should always use ``super().__init__()` instead of ``super(MyNetwork, self).__init__()`` in the PyTorch model, because the latter one has issues with model wrapper.
```
Missing ` at ``super().__init__()`

QuanluZhang replied:
fixed

ultmaster commented on docs/en_US/NAS/ExplorationStrategies.rst:
```diff
@@ -0,0 +1,21 @@
+   * - `TPE Strategy <./ApiReference.rst#nni.retiarii.strategy.TPEStrategy>`__
+     - Sampling new model(s) from user defined model space using `TPE algorithm <https://papers.nips.cc/paper/2011/file/86e8f7ab32cfd12577bc2619bc635690-Paper.pdf>`__ . (``nni.retiarii.strategy.TPEStrategy``)
+   * - `RL Strategy <./ApiReference.rst#nni.retiarii.strategy.PolicyBasedRL>`__
+     - It uses `PPO algorithm <https://arxiv.org/abs/1707.06347>`__ to sample new model(s) from user defined model space. (``nni.retiarii.strategy.PolicyBasedRL``)
```
It's not fair to say so. It supports a series of policy-based RL policies based on tianshou. PPO is used by default.

ultmaster commented on docs/en_US/NAS/MutationPrimitives.rst:
```diff
@@ -0,0 +1,41 @@
+
+* `nn.Repeat <./ApiReference.rst#nni.retiarii.nn.pytorch.Repeat>`__. Repeat a block by a variable number of times.
+
+* `nn.Cell <./ApiReference.rst#nni.retiarii.nn.pytorch.Cell>`__. `This cell structure is popularly used in NAS literature <https://arxiv.org/abs/1611.01578>`__. A cell consists of multiple "nodes". Each node is a sum of multiple operators. Each operator is chosen from ``op_candidates``, and takes one input from previous nodes and predecessors. Predecessor means the input of cell. The output of ce ...
```
We should cite the high-level idea from papers and leave the rest of the technical details to API documentation.

QuanluZhang replied:
I changed some words. we can leave the refactor to the next pr

kvartet commented:
need to update ``toctree`` in the ``docs/en_US/nas.rst``, ``README.md``, and ``docs/en_US/_templates/index.html``.

scarlett2018 commented on docs/en_US/NAS/NasVisualization.rst:
```diff
@@ -0,0 +1,4 @@
+NAS Visualization
+=================
+
+Will be supported soon.
```
add the NAS Visualization issue here so that others knows we are working on it (in which iteration).

scarlett2018 commented on docs/en_US/NAS/Overview.rst:
```diff
@@ -1,5 +1,5 @@
-Neural Architecture Search (NAS) on NNI
-=======================================
+Neural Architecture Search (Retiarii)
```
shall we change to:
**Retiarii for Neural Architecture Search**

as Retiarii is more than NAS.

QuanluZhang replied:
updated

scarlett2018 commented on docs/en_US/NAS/Overview.rst:
```diff
@@ -1,44 +1,59 @@
 
-Automatic neural architecture search is taking an increasingly important role in finding better models. Recent research has proved the feasibility of automatic NAS and has lead to models that beat many manually designed and tuned models. Some representative works are `NASNet <https://arxiv.org/abs/1707.07012>`__\ , `ENAS <https://arxiv.org/abs/1802.03268>`__\ , `DARTS <https://arxiv.org/abs/1806. ...
+There are three key characteristics of the Retiarii framework:
 
```
这段3个特点都挺奇怪的。。我先把好comment的回完，call you to discuss this part....

QuanluZhang replied:
updated based on the discussion

scarlett2018 commented on docs/en_US/NAS/Overview.rst:
```diff
@@ -1,44 +1,59 @@
+There are two types of model space exploration approach: multi-trial NAS and one-shot NAS. Users can construct a model space and use either exploration appraoch to explore the space. Please refer to `Quick Start <./QuickStart.rst>`__ for how to run a NAS experiment on NNI.
 
-With the unified interface, there are two different modes for architecture search. `One <#supported-one-shot-nas-algorithms>`__ is the so-called one-shot NAS, where a super-net is built based on a search space and one-shot training is used to generate a good-performing child model. `The other <#supported-classic-nas-algorithms>`__ is the traditional search-based approach, where each child model w ...
+.. Note:: `The previous NAS framework <https://nni.readthedocs.io/en/v1.9/nas.html>`__ is deprecated, we highly recommend users to move to this Retiarii framework.
```
let's move this up front.

maybe change the tune to： NNI's future NAS supports will all based on Retiarii Framework, users who are still on early version using NNI NAS v1.0 shall migrate your work to Retiarii as soon as possible.

QuanluZhang replied:
updated

scarlett2018 commented on docs/en_US/NAS/Overview.rst:
```diff
@@ -1,44 +1,61 @@
+There are two types of model space exploration approach: **Multi-trial NAS** and **One-shot NAS**. Mutli-trial NAS trains each sampled model in the model space independently, while One-shot NAS samples the model from a super model. After constructing the model space, users can use either exploration appraoch to explore the model space. 
 
-With this motivation, our ambition is to provide a unified architecture in NNI, accelerate innovations on NAS, and apply state-of-the-art algorithms to real-world problems faster.
+Please refer to `Quick Start <./QuickStart.rst>`__ for how to run a NAS experiment on NNI.
```
这个放后面as reference link吧？overview忽然跑出来一个quick start好奇怪。

QuanluZhang replied:
removed

scarlett2018 commented on docs/en_US/NAS/Overview.rst:
```diff
@@ -1,44 +1,61 @@
    :header-rows: 1
    :widths: auto
 
    * - Name
```
change to: Exploration Strategy Name

QuanluZhang replied:
updated

scarlett2018 reviewed (changes requested):
first batch of review comments

scarlett2018 commented on docs/en_US/NAS/OneshotTrainer.rst:
```diff
@@ -0,0 +1,37 @@
+----------------------------------
+
+With a defined model space, users can explore the space in two ways. One is using strategy and single-arch evaluator as demonstrated `here <./Tutorial.rst#explore-the-defined-model-space>`__. The other is using one-shot trainer, which consumes much less computational resource compared to the first one. In this tutorial we focus on this one-shot approach. The principle of one-shot approach is comb ...
+
```
>  It takes charge of both search, training and testing, by training and evaluating this big model.
这句想表达的是啥？

scarlett2018 commented on docs/en_US/NAS/QuickStart.rst:
```diff
@@ -0,0 +1,198 @@
+Quick Start of Retiarii on NNI
+==============================
+
```
another thought is to merge the one-shot approach into 1 quick start which includes:
- define model space
- [A] directly use one-shot trainer
- [B] go with muliti-trial mode: (1) define search strategy; (2) define evaluator;
- start experiment
- view experiment results

scarlett2018 commented on docs/en_US/NAS/QuickStart.rst:
```diff
@@ -0,0 +1,198 @@
+
+**Local Debug Mode**: When running an experiment, it is easy to get some trivial errors in trial code, such as shape mismatch, undefined variable. To quickly fix these kinds of errors, we provide local debug mode which locally applies mutators once and runs only that generated model. To use local debug mode, users can simply invoke the API `debug_mutated_model(base_model, trainer, applied_mutator ...
+
+Visualize the Experiment
```
View experiment results in NNI Web UI

scarlett2018 commented on docs/en_US/NAS/SupermoduleLib.rst:
```diff
@@ -0,0 +1,14 @@
+Super-module Lib
```
what is this file for??

scarlett2018 commented on docs/en_US/NAS/one_shot_nas.rst:
```diff
@@ -1,19 +1,15 @@
+============
 
-One-shot NAS algorithms leverage weight sharing among models in neural architecture search space to train a supernet, and use this supernet to guide the selection of better models. This type of algorihtms greatly reduces computational resource compared to independently training each model from scratch (which we call "Classic NAS"). NNI has supported many popular One-shot NAS algorithms as followi ...
+One-shot NAS algorithms leverage weight sharing among models in neural architecture search space to train a supernet, and use this supernet to guide the selection of better models. This type of algorihtms greatly reduces computational resource compared to independently training each model from scratch (which we call "Multi-trial NAS"). NNI has supported many popular One-shot NAS algorithms as fol ...
```
> This type of algorihtms greatly reduces

algorithms

scarlett2018 commented on docs/en_US/NAS/one_shot_nas.rst:
```diff
@@ -1,19 +1,15 @@
-    FBNet <FBNet>
-    TextNAS <TextNAS>
-    Cream <Cream>
+    Customize one-shot NAS <WriteOneshot>
```
In addition to the out-of-box One-shot algorithms, NNI also provides API for advanced users to define their own one-shot algorithm on NNI. Refer to  **Customize one-shot NAS <WriteOneshot>** for more.

scarlett2018 commented on docs/en_US/NAS/legacy/legacy_nas.rst:
```diff
@@ -0,0 +1,18 @@
+#################
+Legacy NAS on NNI
+#################
```
do we really need a page for this information?

scarlett2018 commented on docs/en_US/NAS/legacy/WriteSearchSpace.rst:
```diff
@@ -0,0 +1,67 @@
+Write A Search Space
+====================
```
shall we get rid of all the legacy contents?

scarlett2018 commented on docs/en_US/NAS/legacy/Visualization.rst:
```diff
@@ -0,0 +1,86 @@
+NAS Visualization (Experimental)
+================================
```
shall we get rid of all the legacy contents?

scarlett2018 commented on docs/en_US/NAS/legacy/TextNAS.rst:
```diff
@@ -0,0 +1,94 @@
+TextNAS
+=======
```
shall we get rid of all the legacy contents?

scarlett2018 commented on docs/en_US/NAS/legacy/SearchSpaceZoo.rst:
```diff
@@ -0,0 +1,281 @@
+.. role:: raw-html(raw)
```
shall we get rid of all the legacy contents?

scarlett2018 commented on docs/en_US/NAS/construct_space.rst:
```diff
@@ -0,0 +1,11 @@
+#####################
```
which file will refer this rst?

scarlett2018 commented on docs/en_US/NAS/WriteTrainer.rst:
```diff
@@ -0,0 +1,96 @@
+Customize A New Model Evaluator
```
trainer equals to evaluator?

scarlett2018 commented on docs/en_US/NAS/WriteTrainer.rst:
```diff
@@ -0,0 +1,96 @@
+Customize A New Model Evaluator
+===============================
+
+Model Evaluator is necessary to evaluate the performance of new explored models. A model evaluator usually includes training, validating and testing of a single model. We provide two ways for users to write a new model evaluator, which will be demonstrated below respectively.
```
我对这个confuse的点比较多，需要call一下~~

QuanluZhang commented on docs/en_US/NAS/Overview.rst:
```diff
@@ -50,67 +67,17 @@ NNI currently supports the one-shot NAS algorithms listed below and is adding mo
-
-* To `report a bug <https://github.com/microsoft/nni/issues/new?template=bug-report.rst>`__ for this feature in GitHub;
-* To `file a feature or improvement request <https://github.com/microsoft/nni/issues/new?template=enhancement.rst>`__ for this feature in GitHub.
+* `Quick Start <./QuickStart.rst>`__ ;
```
added

QuanluZhang commented on docs/en_US/NAS/OneshotTrainer.rst:
```diff
@@ -0,0 +1,37 @@
+One-shot NAS
```
maybe we can leave this refactor to the next pr/release

scarlett2018 commented on docs/en_US/NAS/ExplorationStrategies.rst:
```diff
@@ -0,0 +1,35 @@
+  import nni.retiarii.strategy as strategy
+
+  search_strategy = strategy.Random(dedup=True)  # dedup=False if deduplication is not wanted
+
```
exploration_strategy

QuanluZhang replied:
updated

scarlett2018 commented on docs/en_US/NAS/ExplorationStrategies.rst:
```diff
@@ -0,0 +1,35 @@
+--------------------------------
+
+NNI provides the following exploration strategies for multi-trial NAS. Users could also customize new exploration strategies.
+
```
Add a link to "how to customize a new exploration strategy".

QuanluZhang replied:
added

scarlett2018 commented on docs/en_US/NAS/FBNet.rst:
```diff
@@ -1,6 +1,8 @@
 FBNet
 ======
 
+.. note:: This one-shot NAS is still implemented under NNI NAS 1.0, and will migrated to Retiarii framework in v2.4.
```
add an issue for this migration.

scarlett2018 commented on docs/en_US/nas.rst:
```diff
@@ -1,6 +1,6 @@
+#####################################
+Neural Architecture Search (Retiarii)
+#####################################
 
```
Retiarii for Neural Architecture Search (NAS)

QuanluZhang replied:
updated

kvartet commented on docs/en_US/NAS/OneshotTrainer.rst:
```diff
@@ -0,0 +1,37 @@
+Model Search with One-shot Trainer
+----------------------------------
+
+With a defined model space, users can explore the space in two ways. One is using strategy and single-arch evaluator as demonstrated `here <./Tutorial.rst#explore-the-defined-model-space>`__. The other is using one-shot trainer, which consumes much less computational resource compared to the first one. In this tutorial we focus on this one-shot approach. The principle of one-shot approach is comb ...
```
broken link

QuanluZhang replied:
thanks for finding out the broken link. updated

kvartet reviewed (commented):
some questions:
1. the API reference in the One-shot algorithms doc(like `NAS/ENAS.rst`) is NAS 1.0 API, please update them to the Retiarii API.
2. There is a typo on the first line of the `NAS/Benchmarks.rst`, it should be "improve" instead of "improve", would you mind fixing it in this pr?
3. Shall we remove PPO Tuner in the HPO part in this pr?
4. I saw this [comment](https://github.com/microsoft/nni/issues/2591#issuecomment-857517192), but I didn't find how to do that in the documentation. Did I miss some part?
5. In multi-trial NAS, we can use various training platforms supported by NNI. Do we need to tell users how to use it in this documentation?

QuanluZhang commented on docs/en_US/NAS/ExecutionEngines.rst:
```diff
@@ -0,0 +1,52 @@
+
+1. Add ``@nni.retiarii.model_wrapper`` decorator outside the whole PyTorch model.
+2. If you need to export top models, formatter needs to be set to ``dict``. Exporting ``code`` won't work with this engine.
+
```
removed this point, as pure-python execution engine has become the default one. no need to set formatter any more.
to answer your question, top models can be exported into different format, we can directly export source code of top model, which is ``code`` format. we can also export a dict which records all the chosen values, which is ``dict`` format.

QuanluZhang commented on docs/en_US/NAS/ExecutionEngines.rst:
```diff
@@ -0,0 +1,52 @@
+2. If you need to export top models, formatter needs to be set to ``dict``. Exporting ``code`` won't work with this engine.
+
+.. note:: You should always use ``super().__init__()`` instead of ``super(MyNetwork, self).__init__()`` in the PyTorch model, because the latter one has issues with model wrapper.
+
```
better to add here, because many users may use ``super(MyNetwork, self).__init__()`` in their model. we can also add it in sample code.

QuanluZhang commented on docs/en_US/NAS/ExecutionEngines.rst:
```diff
@@ -0,0 +1,52 @@
+CGO Execution Engine
+--------------------
+
+CGO execution engine does cross-model optimizations based on the graph-based execution engine. This execution engine will be release in v2.4.
```
added

QuanluZhang commented on docs/en_US/nas.rst:
```diff
@@ -10,20 +10,24 @@ However, it takes great efforts to implement NAS algorithms, and it is hard to r
-to accelerate innovations on NAS, and apply state-of-art algorithms on real world problems faster.
-For details, please refer to the following tutorials:
+Thus, we design `Retiarii <https://www.usenix.org/system/files/osdi20-zhang_quanlu.pdf>`__. It is a deep learning framework that supports the exploratory training on a neural network model space, rather than on a single neural network model.
+Exploratory training with Retiarii allows user to express various search space for *Neural Architecture Search* and *Hyper-Parameter Tuning* with high flexibility.
```
updated

QuanluZhang commented on docs/en_US/NAS/QuickStart.rst:
```diff
@@ -197,9 +177,17 @@ After all the above are prepared, it is time to start an experiment to do the mo
 
 The complete code of a simple MNIST example can be found :githublink:`here <examples/nas/multi-trial/mnist/search.py>`.
```
good suggestion, update

QuanluZhang commented on docs/en_US/NAS/MutationPrimitives.rst:
```diff
@@ -0,0 +1,41 @@
+
+To make users easily express a model space within their PyTorch/TensorFlow model, NNI provides some inline mutation APIs as shown below.
+
+* `nn.LayerChoice <./ApiReference.rst#nni.retiarii.nn.pytorch.LayerChoice>`__. It allows users to put several candidate operations (e.g., PyTorch modules), one of them is chosen in each explored model. Note that if the candidate is a user-defined module, it should be decorated as a `basic unit <./Advanced.rst>`__ with ``@basic_unit``. In the following example, ``ops.PoolBN`` and ``ops.SepConv`` s ...
```
good catch! removed this sentence

QuanluZhang commented on docs/en_US/NAS/Mutators.rst:
```diff
@@ -0,0 +1,64 @@
+    stride=stride
+  )
+
+``label`` is used by mutator to identify this placeholder. The other parameters are the information that are required by mutator. They can be accessed from ``node.operation.parameters`` as a dict, it could include any information that users want to put to pass it to user defined mutator. The complete example code can be found in :githublink:`Mnasnet base model <examples/nas/multi-trial/mnasnet/ba ...
```
updated

QuanluZhang commented on docs/en_US/NAS/multi_trial_nas.rst:
```diff
@@ -0,0 +1,13 @@
+Multi-trial NAS
+===============
+
+In multi-trial NAS, users need model evaluator to evaluate the performance of each sampled model, and need an exploration strategy to sample models from a defined model space. Here, users could use NNI provided model evaluators or write their own model evalutor. They can simply choose a exploration strategy. Advanced users can also customize new exploration strategy. For a simple example about ho ...
```
both evaluator and strategy is necessary, let's leave the update in the next pr

QuanluZhang commented on docs/en_US/NAS/ModelEvaluators.rst:
```diff
@@ -0,0 +1,43 @@
+Usage of Model Evaluator
+------------------------
+
+In multi-trial NAS, a sampled model should be able to be executed on a remote machine or a training platform (e.g., AzureML, OpenPAI). Thus, both the model and its model evaluator should be correctly serialized. To make NNI correctly serialize model evaluator, users should apply ``serialize`` on some of their functions and objects
```
updated

kvartet commented on docs/en_US/NAS/ModelEvaluators.rst:
```diff
@@ -0,0 +1,43 @@
+
+NNI provides some commonly used model evaluators for users' convenience. If these model evaluators do not meet users' requirement, they can customize new model evaluators following the tutorial `here <./WriteTrainer.rst>`__.
+
+..  autoclass:: nni.retiarii.evaluator.pytorch.lightning.Classification
```
fix in [this commit](https://github.com/microsoft/nni/pull/3785/commits/a90560f072c6fbedab8f0902b81e676966ecee2c)

QuanluZhang commented:
> some questions:
> 
> 1. the API reference in the One-shot algorithms doc(like `NAS/ENAS.rst`) is NAS 1.0 API, please update them to the Retiarii API.
updated
> 2. There is a typo on the first line of the `NAS/Benchmarks.rst`, it should be "improve" instead of "imporve", would you mind fixing it in this pr?
updated
> 3. Shall we remove PPO Tuner and Network Morphism Tuner in the HPO part in this pr?
removed PPO, keep Network Morphism for now
> 4. I saw this [comment](https://github.com/microsoft/nni/issues/2591#issuecomment-857517192), but I didn't find how to do that in the documentation. Did I miss some part?
replied in this issue
> 5. In multi-trial NAS, we can use various training platforms supported by NNI. Do we need to tell users how to use it in this documentation?
added

kvartet commented:
> need to update `toctree` in the `docs/en_US/nas.rst`, `README.md`, and `docs/en_US/_templates/index.html`.

when updating `README.md`, should update `docs/en_US/_templates/index.html` too, which is the homepage of the readthedocs

kvartet commented:
>  1. the API reference in the One-shot algorithms doc(like `NAS/ENAS.rst`) is NAS 1.0 API, please update them to the Retiarii API.

not just in `ENAS.rst`, the API reference in `DARTS.rst`, `SPOS.rst`, `Proxylessnas.rst` are all outdated.

QuanluZhang commented:
> > 1. the API reference in the One-shot algorithms doc(like `NAS/ENAS.rst`) is NAS 1.0 API, please update them to the Retiarii API.
> 
> not just in `ENAS.rst`, the API reference in `DARTS.rst`, `SPOS.rst`, `Proxylessnas.rst` are all outdated.

updated, thanks for finding out this issue!
